lib/client/kubesession.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/client/terminal"
+	"github.com/gravitational/teleport/lib/kube/proxy/streamproto"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+const mfaChallengeInterval = time.Second * 30
+
+// KubeSession a joined kubernetes session from the client side.
+type KubeSession struct {
+	stream    *streamproto.SessionStream
+	term      *terminal.Terminal
+	ctx       context.Context
+	closeCtx  context.CancelFunc
+	closeWait *sync.WaitGroup
+	meta      types.SessionTracker
+}
+
+// NewKubeSession joins a live kubernetes session.
+func NewKubeSession(ctxBg context.Context, tc *TeleportClient, meta types.SessionTracker, key *Key, kubeAddr string, tlsServer string, mode types.SessionParticipantMode) (*KubeSession, error) {
+	closeWait := &sync.WaitGroup{}
+	joinEndpoint := "wss://" + kubeAddr + "/api/v1/teleport/join/" + meta.GetSessionID()
+	kubeCluster := meta.GetKubeCluster()
+	ciphers := utils.DefaultCipherSuites()
+	tlsConfig, err := key.KubeClientTLSConfig(ciphers, kubeCluster)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if tlsServer != "" {
+		tlsConfig.ServerName = tlsServer
+	}
+
+	dialer := &websocket.Dialer{
+		TLSClientConfig: tlsConfig,
+	}
+
+	ws, resp, err := dialer.Dial(joinEndpoint, nil)
+	if err != nil {
+		body, _ := io.ReadAll(resp.Body)
+		fmt.Printf("Handshake failed with status %d\nand body: %v\n", resp.StatusCode, body)
+		return nil, trace.Wrap(err)
+	}
+
+	stream, err := streamproto.NewSessionStream(ws, streamproto.ClientHandshake{Mode: mode})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	term, err := terminal.New(tc.Stdin, tc.Stdout, tc.Stderr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, closeCtx := context.WithCancel(ctxBg)
+	closeWait.Add(1)
+	go func() {
+		<-ctx.Done()
+		term.Close()
+		closeWait.Done()
+	}()
+
+	if term.IsAttached() {
+		// Put the terminal into raw mode. Note that this must be done before
+		// pipeInOut() as it may replace streams.
+		term.InitRaw(true)
+	}
+
+	handleOutgoingResizeEvent(ctx, stream, term)
+	handleIncomingResizeEvent(stream, closeWait, term)
+	s := &KubeSession{stream, term, ctx, closeCtx, closeWait, meta}
```

Comment:
Just confirming, `KubeSession` does cancel the context at some point, right? If not, you'll leak the resize event goroutine.

Reply:
It does, I've added some extra safeguards and restructured the code to be more clear in that regard.